hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -242,6 +245,17 @@ public int compare(BlockCacheKey a, BlockCacheKey b) {
   /** In-memory bucket size */
   private float memoryFactor;
 
+  private String[] filePaths;
```

Comment:
filePaths applicable for file based BC only.  Now this is made as a member variable in BucketCache itself. Showing a code level bad smell?   At the IOEngine level we have isPersistent() API.  Same way we should have API in IOEngine for generating and verifying the checksum?

Reply:
We add getFilePaths() method in IOEngine, and delete the filePaths variable in BucketCache. Generating and verifying the checksum APIs are in FileIOEngineUtils and FileIOEngine, because only FileIOEngine has to verify the integrity.

Reply:
getFilePaths() in IOEngine?  That seems strange.  The IOEngine need not be file backed always right? Can we have a way in IOEngine to do this checksum calc and write as well as verification within the specific IOE impl class? (Rather than in BucketCache).  We have FileMmapIOE also which also needs this right?

Reply:
FileMmapIOE has not been implemented in branch-1. And if the checksum calculate API in IOEngine level seems strange, because ByteBufferIOE don't have to do this. So we put checksum calc API in a util class, and the specific IOE impl class (FileIOEngine) to do verify.